File: src/NodeParser/TypeofNodeParser.ts
import ts from "typescript";
import type { Context, NodeParser } from "../NodeParser.js";
import type { SubNodeParser } from "../SubNodeParser.js";
import type { BaseType } from "../Type/BaseType.js";
import { ObjectType, ObjectProperty } from "../Type/ObjectType.js";
import type { ReferenceType } from "../Type/ReferenceType.js";
import { getKey } from "../Utils/nodeKey.js";
import { LiteralType } from "../Type/LiteralType.js";
import { NeverType } from "../Type/NeverType.js";
import { FunctionType } from "../Type/FunctionType.js";
import { LogicError } from "../Error/Errors.js";

export class TypeofNodeParser implements SubNodeParser {
    public constructor(
        protected typeChecker: ts.TypeChecker,
        protected childNodeParser: NodeParser,
    ) {}

    public supportsNode(node: ts.TypeQueryNode): boolean {
        return node.kind === ts.SyntaxKind.TypeQuery;
    }

    public createType(node: ts.TypeQueryNode, context: Context, reference?: ReferenceType): BaseType {
        let symbol = this.typeChecker.getSymbolAtLocation(node.exprName)!;
        if (symbol.flags & ts.SymbolFlags.Alias) {
            symbol = this.typeChecker.getAliasedSymbol(symbol);
        }

        const valueDec = symbol.valueDeclaration;
        if (!valueDec) {
            if (symbol.name === "globalThis") {
                // avoids crashes on globalThis but we really shoulodn't try to make a schema for globalThis
                return new NeverType();
            }

            throw new LogicError(node, `No value declaration found for symbol "${symbol.name}"`);
        }

        if (ts.isEnumDeclaration(valueDec)) {
            return this.createObjectFromEnum(valueDec, context, reference);
        }

        if (
            ts.isVariableDeclaration(valueDec) ||
            ts.isPropertySignature(valueDec) ||
            ts.isPropertyDeclaration(valueDec)
        ) {
            let initializer: ts.Expression | undefined;
            if (valueDec.type) {
                return this.childNodeParser.createType(valueDec.type, context);
            }

            if ((initializer = (valueDec as ts.VariableDeclaration | ts.PropertyDeclaration)?.initializer)) {
                return this.childNodeParser.createType(initializer, context);
            }
        }

        if (ts.isClassDeclaration(valueDec)) {
            return this.childNodeParser.createType(valueDec, context);
        }

        if (ts.isPropertyAssignment(valueDec)) {
            return this.childNodeParser.createType(valueDec.initializer, context);
        }

        if (valueDec.kind === ts.SyntaxKind.FunctionDeclaration) {
            return new FunctionType(<ts.FunctionDeclaration>valueDec);
        }

        throw new LogicError(valueDec, `Invalid type query for this declaration. (ts.SyntaxKind = ${valueDec.kind})`);
    }

    protected createObjectFromEnum(node: ts.EnumDeclaration, context: Context, reference?: ReferenceType): ObjectType {
        const id = `typeof-enum-${getKey(node, context)}`;
        if (reference) {
            reference.setId(id);
            reference.setName(id);
        }

        let type: BaseType | null = null;
        const properties = node.members.map((member) => {
            const name = member.name.getText();

            if (member.initializer) {
                type = this.childNodeParser.createType(member.initializer, context);
            } else if (type === null) {
                type = new LiteralType(0);
            } else if (type instanceof LiteralType && typeof type.getValue() === "number") {
                type = new LiteralType(+type.getValue() + 1);
            } else {
                throw new LogicError(member.name, `Enum initializer missing for "${name}"`);
            }

            return new ObjectProperty(name, type, true);
        });

        return new ObjectType(id, [], properties, false);
    }
}

File: internal/parser/typeof_node_parser.go
package parser

import (
	"fmt"
	"strings"

	"github.com/microsoft/typescript-go/shim/ast"
	"github.com/microsoft/typescript-go/shim/checker"
	"github.com/microsoft/typescript-go/shim/scanner"

	"github.com/vega/ts-json-schema-generator/internal/types"
)

// TypeofNodeParser parses `typeof X` type query nodes
// (src/NodeParser/TypeofNodeParser.ts).
type TypeofNodeParser struct {
	typeChecker     *checker.Checker
	childNodeParser NodeParser
}

func NewTypeofNodeParser(typeChecker *checker.Checker, childNodeParser NodeParser) *TypeofNodeParser {
	return &TypeofNodeParser{typeChecker: typeChecker, childNodeParser: childNodeParser}
}

func (p *TypeofNodeParser) SupportsNode(node *ast.Node) bool {
	return node.Kind == ast.KindTypeQuery
}

func (p *TypeofNodeParser) CreateType(node *ast.Node, ctx *Context, reference *types.ReferenceType) types.Type {
	symbol := p.typeChecker.GetSymbolAtLocation(node.AsTypeQueryNode().ExprName)
	if symbol.Flags&ast.SymbolFlagsAlias != 0 {
		symbol = p.typeChecker.GetAliasedSymbol(symbol)
	}

	valueDec := symbol.ValueDeclaration
	if valueDec == nil {
		if symbol.Name == "globalThis" {
			// Avoids crashes on globalThis, but we really shouldn't try to
			// make a schema for globalThis.
			return &types.NeverType{}
		}
		panic(fmt.Errorf("no value declaration found for symbol %q at node %s", symbol.Name, DescribeNode(node)))
	}

	if ast.IsEnumDeclaration(valueDec) {
		return p.createObjectFromEnum(valueDec, ctx, reference)
	}

	if ast.IsVariableDeclaration(valueDec) || ast.IsPropertySignatureDeclaration(valueDec) || ast.IsPropertyDeclaration(valueDec) {
		if declaredType := valueDec.Type(); declaredType != nil {
			return p.childNodeParser.CreateType(declaredType, ctx, nil)
		}
		if initializer := valueDec.Initializer(); initializer != nil {
			return p.childNodeParser.CreateType(initializer, ctx, nil)
		}
	}

	if ast.IsClassDeclaration(valueDec) {
		return p.childNodeParser.CreateType(valueDec, ctx, nil)
	}

	if ast.IsPropertyAssignment(valueDec) {
		return p.childNodeParser.CreateType(valueDec.AsPropertyAssignment().Initializer, ctx, nil)
	}

	if ast.IsFunctionDeclaration(valueDec) {
		return &types.FunctionType{
			Comment:        functionComment(valueDec),
			NamedArguments: namedArguments(p.childNodeParser, valueDec, ctx),
			ReturnType:     returnTypeOfFunction(p.childNodeParser, valueDec, ctx),
		}
	}

	panic(fmt.Errorf("invalid type query for this declaration. (ts.SyntaxKind = %d) at node %s", valueDec.Kind, DescribeNode(valueDec)))
}

func (p *TypeofNodeParser) createObjectFromEnum(node *ast.Node, ctx *Context, reference *types.ReferenceType) types.Type {
	id := "typeof-enum-" + GetNodeKey(node, ctx)
	if reference != nil {
		reference.SetID(id)
		reference.SetName(id)
	}

	var current types.Type
	members := node.Members()
	properties := make([]*types.ObjectProperty, 0, len(members))
	for _, member := range members {
		name := scanner.GetTextOfNode(member.Name())

		if initializer := member.AsEnumMember().Initializer; initializer != nil {
			current = p.childNodeParser.CreateType(initializer, ctx, nil)
		} else if current == nil {
			current = &types.LiteralType{Value: float64(0)}
		} else if literal, isLiteral := current.(*types.LiteralType); isLiteral {
			if value, isNumber := literal.Value.(float64); isNumber {
				current = &types.LiteralType{Value: value + 1}
			} else {
				panic(fmt.Errorf("enum initializer missing for %q at node %s", name, DescribeNode(member.Name())))
			}
		} else {
			panic(fmt.Errorf("enum initializer missing for %q at node %s", name, DescribeNode(member.Name())))
		}

		properties = append(properties, types.NewObjectProperty(name, current, true))
	}

	return types.NewObjectType(id, nil, properties, false, false)
}

// ---------------------------------------------------------------------------
// Helpers shared with FunctionNodeParser/ConstructorNodeParser in the
// TypeScript implementation (src/NodeParser/FunctionNodeParser.ts).

// namedArguments builds the object type describing a function-like node's
// parameters (getNamedArguments in FunctionNodeParser.ts).
func namedArguments(childNodeParser NodeParser, node *ast.Node, ctx *Context) *types.ObjectType {
	parameters := node.Parameters()
	if len(parameters) == 0 {
		return nil
	}

	// Note: the TypeScript implementation returns the parsed InferType
	// directly when the sole parameter is `(...args: infer T)`. Infer types
	// cannot occur in the declarations handled here and the Go FunctionType
	// model only holds an ObjectType, so that special case is omitted.

	properties := make([]*types.ObjectProperty, len(parameters))
	for i, parameter := range parameters {
		parameterType := childNodeParser.CreateType(parameter, ctx, nil)

		// If it's missing a questionToken but has an initializer we can
		// consider the property as not required.
		declaration := parameter.AsParameterDeclaration()
		required := declaration.QuestionToken == nil && declaration.Initializer == nil

		properties[i] = types.NewObjectProperty(scanner.GetTextOfNode(parameter.Name()), parameterType, required)
	}

	return types.NewObjectType("object-"+GetNodeKey(node, ctx), nil, properties, false, false)
}

// returnTypeOfFunction resolves a function-like node's return type
// (getReturnType in FunctionNodeParser.ts).
func returnTypeOfFunction(childNodeParser NodeParser, node *ast.Node, ctx *Context) types.Type {
	returnType := node.Type()
	if returnType == nil {
		return nil
	}
	// Type predicates (`value is T` / `asserts value is T` / `asserts value`)
	// are compile-time-only constructs. At runtime, type guards return boolean
	// and assertion functions return void.
	if ast.IsTypePredicateNode(returnType) {
		if returnType.AsTypePredicateNode().AssertsModifier != nil {
			return &types.VoidType{}
		}
		return &types.BooleanType{}
	}
	return childNodeParser.CreateType(returnType, ctx, nil)
}

// functionComment renders the signature comment stored on FunctionType
// (src/Type/FunctionType.ts constructor), including the "undefined" artifact
// for a missing return type annotation.
func functionComment(node *ast.Node) string {
	source := ast.GetSourceFileOfNode(node)
	fullText := func(n *ast.Node) string {
		return source.Text()[n.Pos():n.End()]
	}

	parameters := node.Parameters()
	parts := make([]string, len(parameters))
	for i, parameter := range parameters {
		parts[i] = fullText(parameter)
	}

	returnText := "undefined"
	if returnType := node.Type(); returnType != nil {
		returnText = fullText(returnType)
	}

	return "(" + strings.Join(parts, ",") + ") =>" + returnText
}
